mods/player_api/api.lua

```diff
@@ -1,54 +1,85 @@
--- Minetest 0.4 mod: player
--- See README.txt for licensing and other information.
-
 player_api = {}
 
 -- Player animation blending
 -- Note: This is currently broken due to a bug in Irrlicht, leave at 0
 local animation_blend = 0
 
-player_api.registered_models = { }
+player_api.registered_models = {}
 
 -- Local for speed.
 local models = player_api.registered_models
 
+local function collisionbox_equals(collisionbox, other_collisionbox)
+	if collisionbox == other_collisionbox then
+		return true
+	end
+	for index = 1, 6 do
+		if collisionbox[index] ~= other_collisionbox[index] then
+			return false
+		end
+	end
+	return true
+end
+
 function player_api.register_model(name, def)
 	models[name] = def
+	def.visual_size = def.visual_size or {x = 1, y = 1}
+	def.collisionbox = def.collisionbox or {-0.3, 0.0, -0.3, 0.3, 1.7, 0.3}
+	def.stepheight = def.stepheight or 0.6
+	def.eye_height = def.eye_height or 1.47
+	for animation_name, animation in pairs(def.animations) do
```

Comment:
This code takes a lot time to read. Please comment the actions done here and why *exactly* the `_equals` copy is needed here. Even after reading the discussion I don't understand how these cannot be taken directly using `table.copy`.

Reply:
I guess you're referring to the loop? It basically finds another animation with equal properties and then takes the key of that animation.

Reply:
The purpose of the loop is to sort animations into "property classes".

All animations in a "property class" have equal properties (equal eye height and collisionbox), so there is no need to call `set_properties` if the previous animation is in the same property class (`previous_anim_equals == anim._equals`).

Those "property classes" use the name of the "first" (in "pairs" traversal order) animation with those properties.

For example, the "property classes" of the default model might be: `{ stand = { ..., _equals = "stand" }, lay = { ..., _equals = "lay" }, walk = { ..., _equals = "stand" }, walk_mine = { ..., _equals = "stand" }, sit = { _equals = "sit" } }`

"stand", "walk" and "walk_mine" all share one _equals value, "sit" and "lay" have their own _equals values